### DISCUSSION TITLE
Recommended pattern to update a "huge" model (ormar + fastapi)

### DISCUSSION THREAD

**Discussion by igormorgado** [collerek/ormar#414] [Q&A] [answered]
Lets say I have a large mode, I will call it here `User`

```
class User(orm.model):
    id: int = orm.Integer(primary_key)
    username: str = orm.String(max_length=256, unique=True, index=True)
    email_address: str = orm.String(max_length=128, unique=True, index=True)
    password: Optional[str] = orm.String(max_length=128, nullable=True)
    first_name: Optional[str] = orm.String(max_length=128, nullable=True)
    last_name: Optional[str] = orm.String(max_length=128, nullable=True)
    # [...] Many other fields
    is_active: bool = orm.Boolean(default=False)
    created_date: datetime.datetime = orm.DateTime(timezone=True, default=datetime.datetime.now(tz=datetime.timezone(datetime.timedelta(hours=-3))))
    updated_date: datetime.datetime = orm.DateTime(timezone=True, default=datetime.datetime.now(tz=datetime.timezone(datetime.timedelta(hours=-3))))
    last_login: Optional[datetime.datetime] = orm.DateTime(nullable=True)

    class Meta(BaseMeta):
        tablename = "users"
```

I have some endpoint to update the user information

```
@app.put('/user/{username}', response_model=User, response_model_exclude={'password'})
async def update_user(username: str, user: User) -> User:
    user_object = User.objects.filter(username=username).get()
    # [...] Update logic here
    return result
```

The user information can be updated with different fields, sometimes user changes its name, or the password, maybe both, all, etc. That is I can't guarantee that I have all fields in any update request, this do not means I need to clean/reset fields not sent in the update request.

Is there any recommended pattern to update the register in this type of scenario, using the  `User`  dictionary sent on body directly, some magical `User.objects.filter(username=username).get().update_magical(**user.dict())` or the way is to iterate over key, values, check each of them and update them accordingly, maybe the new structural pattern matching or the plain old `if elif` over the field names like

```
for key, val in user.dict().items():
    if key == 'username':
        user_object.username = username
    if key == 'password':
        user_object.username = bcrypt.hash(password)
    # [...] logic continues here

await user_object.update()
```

Naturally I can define the method `update_magical()` inside the User class and put all that `if/elif/else` logic there.

Best regards

**Comment by collerek** [comment, marked as answer]
If you refresh or fetch the model before updating you already have all the values as of pre-update state.
That means that you can simply pass all the values you have in your user if they are set and only those values will be updated.
Of course, that poses a problem if you really want to nullify the field, but that would be the same in your if/else scenario.

```python
# in your update function you can simply pass the dict as kwargs
user_object = User.objects.filter(username=username).get()
return user_object.update(**{k: v for k, v in user.dict() if v})
```

If you want, you can further exclude the fields to limit it further (i.e. do not allow password updates regardless if the field is there) you can utilize `_columns` argument of `update` method.

```python
# this will update all set fields except the password even if it's set
user_object = User.objects.filter(username=username).get()
return user_object.update(
         _columns={x for x in User.Meta.model_fields.keys() if x != "password"} ,
         **{k: v for k, v in user.dict() if v}
        )
```

If you need different behaviour based on field you need a custom function to create a final dictionary for your update (i.e. hash the password etc.)
